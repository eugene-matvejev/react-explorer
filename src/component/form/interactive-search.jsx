import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Suggestions from './suggestions';
import InputWithPills from './input-with-pills';
import compose from './compose-form-field';

export const executeOnChange = (props, value) => props.onChange(
    {
        target: {
            getAttribute: (key) => props[key],
            /** required because of PureComponent */
            value: [...value],
        },
    }
);

class InteractiveSearch extends PureComponent {
    constructor({ pattern, isExpanded, options }) {
        super();

        this.state = {
            pattern,
            isExpanded,
            options,
        };

        this.deferredCallback = 0;

        this.onMouseEnter = this.onMouseEnter.bind(this);
        this.onMouseLeave = this.onMouseLeave.bind(this);

        this.onKeyDown = this.onKeyDown.bind(this);

        this.onChange = this.onChange.bind(this);
        this.onRemoveOption = this.onRemoveOption.bind(this);

        this.onSuccess = this.onSuccess.bind(this);
        this.onError = this.onError.bind(this);
    }

    onMouseEnter() {
        this.setState({ isExpanded: true });
    }

    onMouseLeave() {
        this.setState({ isExpanded: false });
    }

    onSuccess(options) {
        const { onSuccess } = this.props;

        this.setState({ options }, () => { onSuccess && onSuccess(this.props, this.state); });
    }

    onError() {
        const { onError } = this.props;

        this.setState({ options: [] }, () => { onError && onError(this.props, this.state); });
    }

    onKeyDown({ target }) {
        if (this.deferredCallback) {
            clearTimeout(this.deferredCallback);
        }

        const { value: pattern } = target;
        const { onFilter, onFilterTimer } = this.props;

        this.setState(
            { pattern },
            () => {
                setTimeout(() => onFilter(this.props, this.state, this.onSuccess, this.onError), onFilterTimer);
            }
        );
    }

    onChange(e) {
        e.preventDefault();
        e.stopPropagation();

        const key = e.target.getAttribute('data-id');
        let { value, maxValues } = this.props;

        if (null === key || value.length >= maxValues) {
            return;
        }

        value.push(this.state.options[key]);

        /** because <FormHandler/> relay on two attributes */
        executeOnChange(this.props, value);
        // this.props.onChange(
        //     {
        //         target: {
        //             getAttribute: (key) => ({
        //                 'data-section': this.props['data-section'],
        //                 'data-field': this.props['data-field'],
        //             })[key],
        //             /** required because of PureComponent */
        //             value: [...value],
        //         },
        //         stopPropagation: () => { },
        //         preventDefault: () => { },
        //     }
        // );
    }

    onRemoveOption(e) {
        e.preventDefault();
        e.stopPropagation();

        const key = e.target.getAttribute('data-id');
        if (null === key) {
            return;
        }
        const { value } = this.props;

        value.splice(key, 1);

        /** because <FormHandler/> relay on two attributes */
        executeOnChange(this.props, value);
        // this.props.onChange(
        //     {
        //         target: {
        //             getAttribute: (key) => ({
        //                 'data-section': this.props['data-section'],
        //                 'data-field': this.props['data-field'],
        //             })[key],
        //             /** required because of PureComponent */
        //             value: [...value],
        //         },
        //         stopPropagation: () => { },
        //         preventDefault: () => { },
        //     }
        // );
    }

    render() {
        const { 'data-cy': cy, className, placeholder, value } = this.props;
        const { isExpanded, options } = this.state;

        return <div
            className={`interactive-search ${className}`}
            onMouseEnter={this.onMouseEnter}
            onMouseLeave={this.onMouseLeave}
        >
            <InputWithPills
                data-cy={cy}
                onFocus={this.onMouseEnter}
                onChange={this.onKeyDown}
                onClick={this.onRemoveOption}
                value={value}
                placeholder={placeholder}
            />
            {
                isExpanded && options && !!options.length
                && <Suggestions
                    data-cy={cy}
                    onClick={this.onChange}
                    options={options}
                    hashmap={value.reduce((acc, { value: v }) => { acc[v] = v; return acc; }, {})}
                />
            }
        </div>;
    }

    static propTypes = {
        'data-cy': PropTypes.string,
        className: PropTypes.string,
        onChange: PropTypes.func.isRequired,
        onFilter: PropTypes.func.isRequired,
        onFilterTimer: PropTypes.number,
        placeholder: PropTypes.string,
        value: PropTypes.arrayOf(
            PropTypes.shape({
                value: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
            })
        ),
        maxValues: PropTypes.number,
        options: PropTypes.arrayOf(
            PropTypes.shape({
                value: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
            })
        ),
        onSuccess: PropTypes.func,
        onError: PropTypes.func,
    }

    static defaultProps = {
        'data-cy': '',
        className: '',
        value: [],
        onFilterTimer: 250,
    }
}

export default compose(InteractiveSearch);
export {
    InteractiveSearch as PureInteractiveSearch,
}
