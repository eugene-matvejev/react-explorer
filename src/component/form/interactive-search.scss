.interactive-search {
    position: relative;

    & > .suggestions {
        position: absolute;
        width: 100%;
        background: white;
    }
}
