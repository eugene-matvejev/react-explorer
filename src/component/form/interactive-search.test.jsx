import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { PureInteractiveSearch as InteractiveSearch, executeOnChange } from './interactive-search';

configure({ adapter: new Adapter() });

describe('<InteractiveSearch/>', () => {
    const props = {
        onChange: () => { },
        onFilter: () => { },
    };

    describe('render', () => {
        it('with default/required props', () => {
            const c = shallow(<InteractiveSearch {...props} />);

            expect(c).toMatchSnapshot();
        });

        describe('with optional props', () => {
            [
                ['data-cy', '{{data-cy}}'],
                ['className', '{{className}}'],
                ['placeholder', '{{placeholder}}'],
                ['maxValues', 1],
                ['onSuccess', () => { }],
                ['onError', () => { }],
                ['value', [{ value: 1, label: '{{label}}' }]],
                ['options', [{ value: 1, label: '{{label}}' }]],
            ].forEach(([prop, v]) => {
                it(`[::${prop}] as "${v}"`, () => {
                    const c = shallow(<InteractiveSearch {...props} {...{ [prop]: v }} />);

                    expect(c).toMatchSnapshot();
                });
            });
        });
    });

    describe('external callbacks', () => {
        describe('::onSuccess', () => {
            it(`should be invoked from internal callback ::onSuccess with component's props and state as payload`, () => {
                const spy = jest.fn();
                const c = shallow(<InteractiveSearch {...props} onSuccess={spy} />);
                c.instance().onSuccess([]);

                expect(spy).toBeCalledWith(c.instance().props, c.state());
            });
        });

        describe('::onError', () => {
            it(`should be invoked from internal callback ::onError with component's props and state as payload`, () => {
                const spy = jest.fn();
                const c = shallow(<InteractiveSearch {...props} onError={spy} />);
                c.instance().onError([]);

                expect(spy).toBeCalledWith(c.instance().props, c.state());
            });
        });

        describe('::onFilter', () => {
            it('should be executed as deferred callback with deferred time from ::onFilterTimer prop', () => {
                jest.useFakeTimers();

                const spy = jest.fn();
                const value = 'v';

                const c = shallow(<InteractiveSearch {...props} onFilter={spy} onFilterTimer={0} />);

                c
                    .find('InputWithPills')
                    .simulate('change', { target: { value } });

                jest.runAllTimers();

                expect(spy).toBeCalledWith(c.instance().props, c.state(), c.instance().onSuccess, c.instance().onError);
            });
        });

        describe('::onChange', () => {
            it('should be invoked from internal callback ::onChange with new value, from a <Suggestions/> click event', () => {
                const spy = spyOn(props, 'onChange');

                shallow(<InteractiveSearch {...props} value={[{ value: 2, label: '2' }]} maxValues={2} />)
                    .setState({ isExpanded: true, options: [{ label: '1', value: 1 }] })
                    .find('Suggestions')
                    .simulate('click', {
                        preventDefault: () => { },
                        stopPropagation: () => { },
                        target: {
                            getAttribute: () => 0,
                        }
                    });

                expect(spy).toBeCalled();
            });

            it('should be invoked from internal callback ::onRemoveOption with new value, from a <InputWithPills/> click event', () => {
                const spy = spyOn(props, 'onChange');

                shallow(<InteractiveSearch {...props} value={[{ value: 2, label: '2' }]} maxValues={2} />)
                    .find('InputWithPills')
                    .simulate('click', {
                        preventDefault: () => { },
                        stopPropagation: () => { },
                        target: {
                            getAttribute: () => 0,
                        }
                    });

                expect(spy).toBeCalled();
            });
        });
    });

    describe('internal callbacks', () => {
        describe('::onMouseEnter', () => {
            it('should be invoked from a "onMouseEnter" event on <div/> wrapper, and set internal state field ::isExpanded to true', () => {
                const spy = spyOn(InteractiveSearch.prototype, 'setState');

                shallow(<InteractiveSearch {...props} />)
                    .find('div[onMouseEnter]')
                    .simulate('mouseEnter');

                expect(spy).toBeCalledWith({ isExpanded: true });
            });
        });

        describe('::onMouseLeave', () => {
            it('should be invoked from a "onMouseLeave" event on <div/> wrapper, and set internal state field ::isExpanded to false', () => {
                const spy = spyOn(InteractiveSearch.prototype, 'setState');

                shallow(<InteractiveSearch {...props} />)
                    .find('div[onMouseLeave]')
                    .simulate('mouseLeave');

                expect(spy).toBeCalledWith({ isExpanded: false });
            });
        });

        describe('::onKeyDown', () => {
            it('should be invoked from a <InputWithPills/> ::onChange event', () => {
                const spy = spyOn(InteractiveSearch.prototype, 'onKeyDown');

                shallow(<InteractiveSearch {...props} />)
                    .find('InputWithPills')
                    .simulate('change');

                expect(spy).toBeCalled();
            });

            it(`should set internal state field ::pattern from payload's ::value field`, () => {
                const spy = spyOn(InteractiveSearch.prototype, 'setState');
                const value = 'v';

                shallow(<InteractiveSearch {...props} />)
                    .find('InputWithPills')
                    .simulate('change', { target: { value } });

                expect(spy).toBeCalledWith({ pattern: value });
            });

            it('should defer external callback ::onFilter execution by time from ::onFilterTimer prop', () => {
                jest.useFakeTimers();

                const spy = jest.fn();
                const value = 'v';

                const c = shallow(<InteractiveSearch {...props} onFilter={spy} />);

                c
                    .find('InputWithPills')
                    .simulate('change', { target: { value } });

                expect(setTimeout).toBeCalledWith(expect.any(Function), c.instance().props['onFilterTimer']);
            });

            it('should clear previously deferred ::onFilter callback, if it been set', () => {
                jest.useFakeTimers();

                const spy = jest.fn();
                const value = 'v';

                const c = shallow(<InteractiveSearch {...props} onFilter={spy} />);

                c.instance().deferredCallback = 1;

                c
                    .find('InputWithPills')
                    .simulate('change', { target: { value } });

                expect(clearTimeout).toBeCalledWith(c.instance().deferredCallback);
            });
        });
    });

    describe('::onChange', () => {
        it('should be invoked from a <Suggestions/> ::onClick event', () => {
            const spy = spyOn(InteractiveSearch.prototype, 'onChange');

            shallow(<InteractiveSearch {...props} />)
                .setState({ isExpanded: true, options: [{ label: '1', value: 1 }] })
                .find('Suggestions')
                .simulate('click');

            expect(spy).toBeCalled();
        });

        it(`should NOT call external callback ::onChange if "data-id" attribute is NOT present in target`, () => {
            const spy = spyOn(props, 'onChange');

            shallow(<InteractiveSearch {...props} />)
                .setState({ isExpanded: true, options: [{ label: '1', value: 1 }] })
                .find('Suggestions')
                .simulate('click', {
                    preventDefault: () => { },
                    stopPropagation: () => { },
                    target: {
                        getAttribute: () => null,
                    }
                });

            expect(spy).not.toBeCalled();
        });

        it(`should NOT call external callback ::onChange if ::value has already max or more options than allowed by ::maxValues prop`, () => {
            const spy = spyOn(props, 'onChange');

            shallow(<InteractiveSearch {...props} value={[{ value: 1, label: '1' }]} maxValues={1} />)
                .setState({ isExpanded: true, options: [{ label: '1', value: 1 }] })
                .find('Suggestions')
                .simulate('click', {
                    preventDefault: () => { },
                    stopPropagation: () => { },
                    target: {
                        getAttribute: () => 1,
                    }
                });

            expect(spy).not.toBeCalled();
        });
    });

    describe('::onRemoveOption', () => {
        it('should be invoked from a <InputWithPills/> ::onClick event', () => {
            const spy = spyOn(InteractiveSearch.prototype, 'onRemoveOption');

            shallow(<InteractiveSearch {...props} />)
                .find('InputWithPills')
                .simulate('click');

            expect(spy).toBeCalled();
        });

        it(`should NOT call external callback ::onChange if "data-id" attribute is NOT present in target`, () => {
            const spy = spyOn(props, 'onChange');

            shallow(<InteractiveSearch {...props} />)
                // .setState({ isExpanded: true })
                .find('InputWithPills')
                .simulate('click', {
                    preventDefault: () => { },
                    stopPropagation: () => { },
                    target: {
                        getAttribute: () => null,
                    }
                });

            expect(spy).not.toBeCalled();
        });
    });
});

describe('on change proxy', () => {
    it(`should mock React like HTML execution, with 'data-field', and 'data-section' attributes`, () => {
        let _;

        const props = {
            onChange: (v) => { _ = v},
            'data-field': 0,
            'data-section': 0,
        };

        executeOnChange(props, []);

        _.stopPropagation();
        _.preventDefault();

        expect(_.target.getAttribute('data-field')).toBe(0);
        expect(_.target.getAttribute('data-section')).toBe(0);
    });
})
