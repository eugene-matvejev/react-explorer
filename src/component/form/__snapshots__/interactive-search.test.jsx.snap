// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InteractiveSearch/> render with default/required props 1`] = `
<div
  className="interactive-search "
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <InputWithPills
    data-cy=""
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    value={Array []}
  />
</div>
`;

exports[`<InteractiveSearch/> render with optional props [::className] as "{{className}}" 1`] = `
<div
  className="interactive-search {{className}}"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <InputWithPills
    data-cy=""
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    value={Array []}
  />
</div>
`;

exports[`<InteractiveSearch/> render with optional props [::data-cy] as "{{data-cy}}" 1`] = `
<div
  className="interactive-search "
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <InputWithPills
    data-cy="{{data-cy}}"
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    value={Array []}
  />
</div>
`;

exports[`<InteractiveSearch/> render with optional props [::maxValues] as "1" 1`] = `
<div
  className="interactive-search "
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <InputWithPills
    data-cy=""
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    value={Array []}
  />
</div>
`;

exports[`<InteractiveSearch/> render with optional props [::onError] as "() => {}" 1`] = `
<div
  className="interactive-search "
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <InputWithPills
    data-cy=""
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    value={Array []}
  />
</div>
`;

exports[`<InteractiveSearch/> render with optional props [::onSuccess] as "() => {}" 1`] = `
<div
  className="interactive-search "
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <InputWithPills
    data-cy=""
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    value={Array []}
  />
</div>
`;

exports[`<InteractiveSearch/> render with optional props [::options] as "[object Object]" 1`] = `
<div
  className="interactive-search "
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <InputWithPills
    data-cy=""
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    value={Array []}
  />
</div>
`;

exports[`<InteractiveSearch/> render with optional props [::placeholder] as "{{placeholder}}" 1`] = `
<div
  className="interactive-search "
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <InputWithPills
    data-cy=""
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    placeholder="{{placeholder}}"
    value={Array []}
  />
</div>
`;

exports[`<InteractiveSearch/> render with optional props [::value] as "[object Object]" 1`] = `
<div
  className="interactive-search "
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <InputWithPills
    data-cy=""
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    value={
      Array [
        Object {
          "label": "{{label}}",
          "value": 1,
        },
      ]
    }
  />
</div>
`;
